amp/shard_tracker.go

```diff
@@ -0,0 +1,161 @@
+	child *Child
+	mpp   *record.MPP
+	amp   *record.AMP
+}
+
+// A compile time check to ensure Shard implements the shards.PaymentShard
+// interface.
+var _ shards.PaymentShard = (*Shard)(nil)
+
+// Hash returns the hash used for the HTLC representing this AMP shard.
+func (s *Shard) Hash() lntypes.Hash {
+	return s.child.Hash
+}
+
+// MPP returns any extra MPP records that should be set for the final hop on
+// the route used by this shard.
+func (s *Shard) MPP() *record.MPP {
+	return s.mpp
+}
+
+// AMP returns any extra AMP records that should be set for the final hop on
+// the route used by this shard.
+func (s *Shard) AMP() *record.AMP {
+	return s.amp
+}
+
+// ShardTracker is an implementation of the shards.ShardTracker interface
+// that is able to generate payment shards according to the AMP splitting
+// algorithm. It can be used to generate new hashes to use for HTLCs, and also
+// cancel shares used for failed payment shards.
+type ShardTracker struct {
+	setID       [32]byte
+	paymentAddr [32]byte
+	totalAmt    lnwire.MilliSatoshi
+
+	sharer *SeedSharer
+
+	shards map[uint64]*Child
+}
+
+// A compile time check to ensure ShardTracker implements the
+// shards.ShardTracker interface.
+var _ shards.ShardTracker = (*ShardTracker)(nil)
+
+// NewShardTracker creates a new shard tracker to use for AMP payments. The
+// root shard, setID, payment address and total amount must be correctly set in
+// order for the TLV options to include with each shard to be created
+// correctly.
+func NewShardTracker(root, setID, payAddr [32]byte,
+	totalAmt lnwire.MilliSatoshi) *ShardTracker {
+
+	// Create a new seed sharer from this root.
+	rootShare := Share(root)
+	rootSharer := SeedSharerFromRoot(&rootShare)
+
+	return &ShardTracker{
+		setID:       setID,
+		paymentAddr: payAddr,
+		totalAmt:    totalAmt,
+		sharer:      rootSharer,
+		shards:      make(map[uint64]*Child),
+	}
+}
+
+var zeroShare = Share{}
+
+// NewShard registers a new attempt with the ShardTracker and returns a
+// new shard representing this attempt. This attempt's shard should be canceled
+// if it ends up not being used by the overall payment, i.e. if the attempt
+// fails.
+func (s *ShardTracker) NewShard(pid uint64, last bool) (shards.PaymentShard,
+	error) {
+
+	// If the current share is a zero share, it means we already returned
+	// the last shard.
+	if s.sharer.curr == zeroShare {
+		return nil, fmt.Errorf("splitter exhausted")
+	}
+
+	// Use a random child index.
+	var childIndex [4]byte
+	if _, err := rand.Read(childIndex[:]); err != nil {
+		return nil, err
+	}
+	idx := binary.BigEndian.Uint32(childIndex[:])
+
+	// Depending on whether we are requesting the last shard or not, either
+	// split the current share into two, or return it.
+	var child *Child
+	if last {
+		child = s.sharer.Child(idx)
+
+		// If this was the last shard, set the current share to the
+		// zero share to indicate we cannot split it further.
+		s.sharer.curr = zeroShare
+	} else {
```

Comment:
When it comes to the splitting, don't we only need to re-share when we split the greater payment? So if we need a new shard to retry a value of that same amount, we can just increment the prior index used.

Reply:
I guess there's a good bit of freedom w.r.t to implementation here in the end. What we have right now will keep on adding a new layer to the tree (and essentially removing the stale ones) when we need to retry rather than expanding to select different children for a given leaf node.

Reply:
> When it comes to the splitting, don't we only need to re-share when we split the greater payment?

This is true, but it is more generic to always `Merge->Split` when we retry, since we need to do this anyway in the case where the pathfinder decides to change the value.